# TikTok Project

**The purpose** of this project is to investigate and understand the data provided. This activity will:
1. Acquaint you with the data
2. Compile summary information about the data
3. Begin the process of EDA and reveal insights contained in the data
4. Prepare you for more in-depth EDA, hypothesis testing, and statistical analysis

**The goal** is to construct a dataframe in Python, perform a cursory inspection of the provided dataset, and inform TikTok data team members of your findings. <br>
<i>This activity has three parts</i>:
<br><br>
**Part 1**: Understand the situation
* How can I best prepare to understand and organize the provided TikTok information?

**Part 2**: Understand the data
*  Create a pandas dataframe for data learning and future exploratory data analysis (EDA) and statistical activities
*  Compile summary information about the data to inform next steps

**Part 3**: Understand the variables
* Use insigts from my examination of the summary data to guide deeper investingation into variables.

<a id=Top></a>
## Idendify data types and compile summary information
USE the problem-solving framework PACE:<br>
### PACE stages
[<u>Plan</u>](#1)<br>
[<u>Analyze</u>](#2)<br>
[<u>Construct</u>](#3)<br>
[<u>Execute</u>](#4)

<a id = 1></a>
### PACE : Plan
#### Task1. Understand the situation
Read, explore, and identify key variables using the provided csv files.

<a id = 2></a>
### PACE : Analyze
#### Task2a. Imports and data loading

In [41]:
import pandas as pd
import numpy as np

data = pd.read_csv("tiktok_dataset.csv", encoding='cp1252')

#### Taks2b. Understand the data - Inspect the data

In [42]:
data.head(10)

# claim_status    video_id  video_duration_sec  \
0   1        claim  7017666017                  59   
1   2        claim  4014381136                  32   
2   3        claim  9859838091                  31   
3   4        claim  1866847991                  25   
4   5        claim  7105231098                  19   
5   6        claim  8972200955                  35   
6   7        claim  4958886992                  16   
7   8        claim  2270982263                  41   
8   9        claim  5235769692                  50   
9  10        claim  4660861094                  45   

                            video_transcription_text verified_status  \
0  someone shared with me that drone deliveries a...    not verified   
1  someone shared with me that there are more mic...    not verified   
2  someone shared with me that american industria...    not verified   
3  someone shared with me that the metro of st. p...    not verified   
4  someone shared with me that the number of busi...    not verified   
5  someone shared with me that gross domestic pro...    not verified   
6  someone shared with me that elvis presley has ...    not verified   
7  someone shared with me that the best selling s...    not verified   
8  someone shared with me that about half of the ...    not verified   
9  someone shared with me that it would take a 50...        verified   

  author_ban_status  video_view_count  video_like_count  video_share_count  \
0      under review          343296.0           19425.0              241.0   
1            active          140877.0           77355.0            19034.0   
2            active          902185.0           97690.0             2858.0   
3            active          437506.0          239954.0            34812.0   
4            active           56167.0           34987.0             4110.0   
5      under review          336647.0          175546.0            62303.0   
6            active          750345.0          486192.0           193911.0   
7            active          547532.0            1072.0               50.0   
8            active           24819.0           10160.0             1050.0   
9            active          931587.0          171051.0            67739.0   

   video_download_count  video_comment_count  
0                   1.0                  0.0  
1                1161.0                684.0  
2                 833.0                329.0  
3                1234.0                584.0  
4                 547.0                152.0  
5                4293.0               1857.0  
6                8616.0               5446.0  
7                  22.0                 11.0  
8                  53.0                 27.0  
9                4104.0               2540.0

In [43]:
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 19382 entries, 0 to 19381
Data columns (total 12 columns):
 #   Column                    Non-Null Count  Dtype  
---  ------                    --------------  -----  
 0   #                         19382 non-null  int64  
 1   claim_status              19084 non-null  object 
 2   video_id                  19382 non-null  int64  
 3   video_duration_sec        19382 non-null  int64  
 4   video_transcription_text  19084 non-null  object 
 5   verified_status           19382 non-null  object 
 6   author_ban_status         19309 non-null  object 
 7   video_view_count          19084 non-null  float64
 8   video_like_count          19084 non-null  float64
 9   video_share_count         19084 non-null  float64
 10  video_download_count      19084 non-null  float64
 11  video_comment_count       19084 non-null  float64
dtypes: float64(5), int64(3), object(4)
memory usage: 1.8+ MB


In [44]:
data.describe()

#      video_id  video_duration_sec  video_view_count  \
count  19382.000000  1.938200e+04        19382.000000      19084.000000   
mean    9691.500000  5.627454e+09           32.421732     254708.558688   
std     5595.245794  2.536440e+09           16.229967     322893.280814   
min        1.000000  1.234959e+09            5.000000         20.000000   
25%     4846.250000  3.430417e+09           18.000000       4942.500000   
50%     9691.500000  5.618664e+09           32.000000       9954.500000   
75%    14536.750000  7.843960e+09           47.000000     504327.000000   
max    19382.000000  9.999873e+09           60.000000     999817.000000   

       video_like_count  video_share_count  video_download_count  \
count      19084.000000       19084.000000          19084.000000   
mean       84304.636030       16735.248323           1049.429627   
std       133420.546814       32036.174350           2004.299894   
min            0.000000           0.000000              0.000000   
25%          810.750000         115.000000              7.000000   
50%         3403.500000         717.000000             46.000000   
75%       125020.000000       18222.000000           1156.250000   
max       657830.000000      256130.000000          14994.000000   

       video_comment_count  
count         19084.000000  
mean            349.312146  
std             799.638865  
min               0.000000  
25%               1.000000  
50%               9.000000  
75%             292.000000  
max            9599.000000

*Consider the following questions:*

**Question 1:** When reviewing the first few rows of the dataframe, what do you observe about the data? What does each row represen
* each row represents unique viedo and own attributest?

**Question 2:** When reviewing the `data.info()` output, what do you notice about the different variables? Are there any null values? Are all of the variables numeric? Does anything else stand 

* No NULL value, Dtypes are 5 floats, 3 int and 4 object.out?

**Question 3:** When reviewing the `data.describe()` output, what do you notice about the distributions of each variable? Are there any questionable values? Does it seem that there are outlier v

* Quite wide std range for video_view_count, video_like_countalues?

#### Task2c. Understand the data - Investigate the variables

In [45]:
# What are the different values for claim status and how many of each are in the data?
data.claim_status.value_counts()

claim_status
claim      9608
opinion    9476
Name: count, dtype: int64

* The count of claim status are quite balanced.

Next, examine the engagement trends associated with each different claim status.

In [60]:
# What is the average view count of videos with "claim" status?
avg_view_count = data[data['claim_status'] == 'claim']
print(f"Avg. view count of videos with 'claim': {avg_view_count.video_view_count.mean()}")
print(f"Median view count of videos with 'claim': {avg_view_count.video_view_count.median()}")

Avg. view count of videos with 'claim': 501029.4527477102
Median view count of videos with 'claim': 501555.0


In [61]:
# What is the average view count of videos with "opinion" status?
avg_view_count = data[data['claim_status'] == 'opinion']
print(f"Avg. view count of videos with 'opinion': {avg_view_count.video_view_count.mean()}")
print(f"Median view count of videos with 'claim': {avg_view_count.video_view_count.median()}")

Avg. view count of videos with 'opinion': 4956.43224989447
Median view count of videos with 'claim': 4953.0


* The mean and median within each claim category are close to one another, but there is a vast discrepancy between view count for videos labeled as claims and videos labeled as opinons.

Now, examine trends associated with the ban status of the author.

In [65]:
# Get counts for each group combination of claim status and author ban status
# data.groupby(['claim_status', 'author_ban_status']).count()[['#']]
data.groupby(['claim_status', 'author_ban_status']).size()

claim_status  author_ban_status
claim         active               6544
              banned               1436
              under review         1592
opinion       active               8782
              banned                196
              under review          461
dtype: int64

* There's a noticeable difference in the number of banned authors between 'claim' and 'opinion' video. 1439 with authors that are banned. otherwise, 196 have banned authors that are opinion.
    * Claim video are more strictly policed than opnion videos
    * Authors must comply with a stricter set of rules if they post a claim than if they post an opinion.

* The video from banned authors have most significant values compare others.

In [68]:
data.groupby(['author_ban_status']).agg({'video_view_count':['mean','median'],
                                         'video_like_count':['mean','median'],
                                         'video_share_count':['mean', 'median']
                                        })

video_view_count           video_like_count            \
                              mean    median             mean    median   
author_ban_status                                                         
active               215873.413285    8623.5     71056.116469    2222.0   
banned               445579.695466  448302.5    153074.835784  105538.0   
under review         391955.077448  364295.0    128530.380419   70805.0   

                  video_share_count           
                               mean   median  
author_ban_status                             
active                 14114.923724    437.0  
banned                 30035.354779  14484.5  
under review           25677.130541   9381.0

In [72]:
# What's the median video share count of each author ban status?
data.groupby(['author_ban_status']).median(numeric_only=True)[['video_share_count']]

video_share_count
author_ban_status                   
active                         437.0
banned                       14484.5
under review                  9381.0

* Banned author have a median share count that's 33 times the median share count of active authors! Explore this in more depth.

Group the data by `author_ban_status` with mean, median and count of `video_view_count`, `video_like_count` and `video_share_count` columns.

In [74]:
# authore_bans_status - count, mean, median of each of following columns.
data.groupby(['author_ban_status']).agg({'video_view_count':['count','mean','median'],
                                         'video_like_count':['count','mean','median'],
                                         'video_share_count':['count','mean','median']})

video_view_count                          video_like_count  \
                             count           mean    median            count   
author_ban_status                                                              
active                       15326  215873.413285    8623.5            15326   
banned                        1632  445579.695466  448302.5             1632   
under review                  2053  391955.077448  364295.0             2053   

                                           video_share_count                \
                            mean    median             count          mean   
author_ban_status                                                            
active              71056.116469    2222.0             15326  14114.923724   
banned             153074.835784  105538.0              1632  30035.354779   
under review       128530.380419   70805.0              2053  25677.130541   

                            
                    median  
author_ban_status           
active               437.0  
banned             14484.5  
under review        9381.0

* A few observations figure out:
  * Banned authors and those under review get far more views, like, and shares than active authors.
  * Across all ban status, most mean values much higher than median except view count.

In [75]:
# Create a likes_per_view column
data['likes_per_view'] = data['video_like_count']/data['video_view_count']
# Create a comments_per_view column
data['comments_per_view'] = data['video_comment_count'] / data['video_view_count']
# Create a shares_per_view column
data['shares_per_view'] = data['video_share_count']/data['video_view_count']

In [77]:
data.groupby(['claim_status', 'author_ban_status']).agg(
    {'likes_per_view':['count','mean','median'],
     'comments_per_view' : ['count','mean','median'],
     'shares_per_view' : ['count','mean','median']
    }
)

likes_per_view                      \
                                        count      mean    median   
claim_status author_ban_status                                      
claim        active                      6544  0.329674  0.327019   
             banned                      1436  0.345245  0.359394   
             under review                1592  0.327864  0.320946   
opinion      active                      8782  0.219600  0.217830   
             banned                       196  0.206868  0.198483   
             under review                 461  0.226549  0.228051   

                               comments_per_view                      \
                                           count      mean    median   
claim_status author_ban_status                                         
claim        active                         6544  0.001394  0.000776   
             banned                         1436  0.001375  0.000742   
             under review                   1592  0.001365  0.000789   
opinion      active                         8782  0.000517  0.000251   
             banned                          196  0.000434  0.000193   
             under review                    461  0.000538  0.000296   

                               shares_per_view                      
                                         count      mean    median  
claim_status author_ban_status                                      
claim        active                       6544  0.065478  0.049315  
             banned                       1436  0.068004  0.051879  
             under review                 1592  0.065611  0.049777  
opinion      active                       8782  0.043707  0.032384  
             banned                        196  0.040531  0.030728  
             under review                  461  0.044410  0.035027

* We knew that videos by banned and under reviewed got far more views, likes and shares than active authors. However, as a view from engagement rates, the claim status is more related than ban status. 
* Both claim and opinion, most are in an active states however it's notable that in the case of opinion is far more active count and far less than comments.. almost no comments.

<a id = 3></a>
### PACE : Construct
**Note**: The Construct stage does not apply to this workflow. The PACE framework can be adapted to fit the specific requirements of any product

<a id = 4></a>
### PACE : Execute
###### Given the efforts, what can I summarize for Rosie Mae Bradshaw and the TikTok data team?
* Total 19382 samples in the dataset, almost 50% are claims ; 9,608 cases of them.
* Engagement level is strongly correlated with claim status. This should be a focus of further inquiry.
* Videos with banned authors have significantly higher engagement than videos with active authors.

[<u><i>Go to Top</i></u>](#Top)